src/cmd/linuxkit/moby/output.go

```diff
@@ -362,6 +375,21 @@ func outputIso(image, filename string, filesystem io.Reader) error {
 	return dockerRun(filesystem, output, image)
 }
 
+func outputIsoInitrd(image, filename string, kernel []byte, initrd []byte, cmdline string) error {
```

Comment:
> Will be addressed in the commit message shortly. Let me know if it's not OK.

My concern is not so much the change log / commit history. It is more users who come to it and have no idea what the difference is between the two formats, and which to choose.

Beyond that, you had suggested that the original iso-efi might not be needed at all. Can you elaborate on that? If we _really_ don't need it anymore, then we can just call this `iso-efi` and replace the other. But we should understand if that is true and be sure.

Reply:
> Beyond that, you had suggested that the original iso-efi might not be needed at all. Can you elaborate on that? If we _really_ don't need it anymore, then we can just call this `iso-efi` and replace the other. But we should understand if that is true and be sure.

Well I said that based on my use cases and on those I can foresee, which are far from the whole thing. For one, there are probably people who use (and depend on) `iso-efi`, so IMHO that's already enough reason to not remove it (at least not _right now_, that is). I'm happy to learn of use cases and scenarios I haven't thought of. For _my_ scenarios, I'm sure I'll only use `iso-efi-initrd` from now on.

Reply:
> Well I said that based on my use cases and on those I can foresee, which are far from the whole thing.

haha! You likely are correct. OK, let us leave it separate for now.

Reply:
> Yeah, I think the log message is fine. If you _really_ care, you can make the type a string parameter to the func, so it will know what to put in the log message. I think that is overkill, but if it works for you, that is fine, too.

Done.

Reply:
> It's the same with the currently supported formats, isn't it? I had to look at the code to understand what each does.

Somewhat. I admit quite freely we need better documentation on the output formats. But most of them are moderately self-explanatory: `iso-efi` is a bootable ISO image with EFI boot structure; `kernel+initrd` outputs a kernel and initrd; `qcow2-bios` as a qcow2 image with BIOS boot structure; etc.

I am not quite clear what `iso-efi-initrd` is and how it differs.

I _think_ (but am unsure) that:

* `iso-efi` is an ISO image with a kernel on it, essentially it runs the `bootx86.efi` (adapt per arch), which loads the kernel and then mounts the ISO read-only as the filesystem
* `iso-efi-initrd` is an ISO image with a kernel on it, essentially it runs the `bootx86.efi` (adapt per arch), loads the kernel and initrd into memory and run

So the only thing on the ISO in this case is the initrd? As opposed to a very small initrd whose purpose is just to mount the ISO root filesystem?

More practically, if my linuxkit-built image has 1000 files in its root. `iso-efi` will add the kernel and a very small initrd, and on boot:

1. efi
2. kernel
3. minimal initrd
4. iso with 1000 files as root filesystem

In the `iso-efi-initrd` case:

1. efi
2. kernel
3. initrd with 1000 files in memory

If that is correct, then the difference certainly will be memory usage, so we would need both `iso-efi` and `iso-efi-initrd`. But I am not sure I got it correctly.

Reply:
> More practically, if my linuxkit-built image has 1000 files in its root. iso-efi will add the kernel and a very small initrd,

`iso-efi` AFAICT adds no initrd, not even a small one. The kernel itself mounts the root filesystem using hardware drivers (which can be a problem for the reasons explained), then runs `/sbin/init` directly.

> If that is correct, then the difference certainly will be memory usage, so we would need both `iso-efi` and `iso-efi-initrd`. But I am not sure I got it correctly.

Yes, that's basically it. Note that the 1000 files problem applies equally to raw, and any other formats that pack the full filesystem into the initrd (efi-initrd, qcow2, vmdk, etc), so it's something that's already being done.

I'm not claiming `iso-efi-initrd` it's the perfect soution (although _for me_ it is); it's a fair tradeoff worth having IMHO.

Reply:
> iso-efi AFAICT adds no initrd, not even a small one. The kernel itself mounts the root filesystem using hardware drivers (which can be a problem for the reasons explained), then runs /sbin/init directly

Good point. I have not looked at the source to that in quite some time.

These could also be solved with a small initrd.

> Yes, that's basically it. Note that the 1000 files problem applies equally to raw, and any other formats that pack the full filesystem into the initrd (efi-initrd, qcow2, vmdk, etc), so it's something that's already being done.

Agreed. But it is a reason we have had in the past _not_ to just have those.

> If one prefers iso-efi with its shortcomings (which might not apply in their case), they can still use it.

Then we keep both.

Reply:
Hold on a sec, don't I still have the chicken-and-egg problem? If I have an issue in some cases with mounting the iso as root, don't I also have issues mounting it to get the initrd? Or the kernel for that matter?

Reply:
> These could also be solved with a small initrd.

Sure, but it requires substantial code refactoring to accomodate for what would be the first instance of something like that; on the other hand, `iso-efi-initrd` can be added here and now by reusing existing code and adding only a tiny builder.

Reply:
Good point. I think what we need, then, is 2 iso-efi (and raw-efi, and etc) modes:

1. root disk distributed on the ISO, kernel and minimal initrd. kernel -> initrd -> mount root (modify existing iso-efi)
2. root disk entirely in initrd (what you are proposing here). kernel -> initrd (done)

BTW, I do find it interesting that the mounting is an issue. Doesn't UEFI support drivers, including in the ESP?